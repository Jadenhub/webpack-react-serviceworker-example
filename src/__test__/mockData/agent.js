export const MOBILE_AGENT = 'Mozilla/5.0 (iPhone; CPU iPhone OS 7_1_2 like Mac OS X) AppleWebKit/537.51.2 (KHTML, like Gecko) OPiOS/10.2.0.93022 Mobile/11D257 Safari/9537.53'
export const DESKTOP_AGENT = 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/42.0.2311.135 Safari/537.36 Edge/12.246'