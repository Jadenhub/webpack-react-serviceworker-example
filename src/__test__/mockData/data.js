export const RATES = {
  "rates":{
    "btc": {
      "name":"Bitcoin",
      "unit":"BTC",
      "value":1.0,
      "type":"crypto"
    },
    "eth": {
      "name":"Ether",
      "unit":"ETH",
      "value":16.198,
      "type":"crypto"
    },
    "ltc": {
      "name":"Litecoin",
      "unit":"LTC",
      "value":230.349,
      "type":"crypto"
    },
    "bch": {
      "name":"Bitcoin Cash",
      "unit":"BCH",
      "value":64.588,
      "type":"crypto"
    },
    "bnb": {
      "name":"Binance Coin",
      "unit":"BNB",
      "value":105.539,
      "type":"crypto"
    },
    "eos": {
      "name":"EOS",
      "unit":"EOS",
      "value":8003.441,
      "type":"crypto"
    },
    "xrp": {
      "name":"XRP",
      "unit":"XRP",
      "value":47106.088,
      "type":"crypto"
    },
    "xlm": {
      "name":"Lumens",
      "unit":"XLM",
      "value":119836.321,
      "type":"crypto"
    },
    "link": {
      "name":"Chainlink",
      "unit":"LINK",
      "value":1693.022,
      "type":"crypto"
    },
    "dot":{
      "name":"Polkadot",
      "unit":"DOT",
      "value":1736.006,
      "type":"crypto"
    },
    "yfi":{
      "name":"Yearn.finance",
      "unit":"YFI",
      "value":0.985,
      "type":"crypto"
    }
  }
};
