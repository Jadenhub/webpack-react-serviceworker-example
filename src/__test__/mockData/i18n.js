export const I18N = {
  PROJECT_NAME: 'webapp',
  META_TITLE: 'Webapp',
  RATES: 'Rates',
  RATES_LIST: 'RATES LIST',
  NAME: 'Name',
  TYPE: 'Type',
  VALUE: 'Value',
  UNIT: 'Unit',
  SELECT_ALL: 'Select All',
  PAGE: 'Page',
  GO_TO_PAGE: 'Go to page',
  SHOW: 'Show',
  OF: 'of'
}