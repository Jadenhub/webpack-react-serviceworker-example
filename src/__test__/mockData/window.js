export function MockWindow(){
  global.window = {
    addEventListener: jest.fn(),
    removeEventListener: jest.fn()
  }
}