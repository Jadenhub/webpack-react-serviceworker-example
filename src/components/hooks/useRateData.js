import { useMemo } from "react";

export function getTableData(rates){
  return Object.keys(rates).map(key=>{
    return rates[key];
  })
}

export function useRateData(rates){
  return useMemo(() => getTableData(rates), [rates])
}