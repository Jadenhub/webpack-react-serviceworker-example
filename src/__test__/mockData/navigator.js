import { DESKTOP_AGENT, MOBILE_AGENT } from "./agent";


export function MockDesktopNavigator(){
  global.navigator = {
    userAgent: DESKTOP_AGENT
  }
}

export function MockMobileNavigator(){
  global.navigator = {
    userAgent: MOBILE_AGENT
  }
}