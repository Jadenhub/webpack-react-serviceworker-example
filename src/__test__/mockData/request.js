import { DESKTOP_AGENT, MOBILE_AGENT } from "./agent"

export function getMockDesktopRequest(){
  return {
    path: '/',
    headers: {
      'user-agent': DESKTOP_AGENT
    }
  }
}

export function getMockMobileRequest(){
  return {
    path: '/',
    headers: {
      'user-agent': MOBILE_AGENT
    }
  }
}