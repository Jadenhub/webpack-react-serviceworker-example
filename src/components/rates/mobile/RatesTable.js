import React, { useEffect, useState } from 'react';
import styled from 'styled-components';
import { withI18n } from '../../decorator/i18n';
import Item from './Item';
import Button from '../../interact/Button';
import InfiniteScroll from '../../common/InfiniteScroll';
import DelayedUiWrapper from '../../style/DelayedUiWrapper';

const RATE_COUNT = 10;

const UlWrapper = styled.ul`
  list-style: none;
  padding: 0;
`;
const LiWrapper = styled.li`
  margin-bottom:10px;
`;

const RatesTableWrapper = styled.div`
  margin: 20px 0 60px 0;
`

function getItem({rates, itemsClicked, setItemsClicked}){
  return rates.map((rate, idx) => {
    return (
      <LiWrapper key={idx} >
        <Item 
          clicked={itemsClicked[idx]}
          rate={rate} 
          onClick={()=>{
            itemsClicked = [...itemsClicked];
            itemsClicked[idx] = !itemsClicked[idx]
            setItemsClicked(itemsClicked)
          }}
        />
      </LiWrapper>
    )
  })
}

function onButtonClick({
  data,
  selectedAll,
  setItemsClicked,
  setSelectedAll
}){
  if (selectedAll) {
    setItemsClicked(Array(data.length).fill(false))
  } else {
    setItemsClicked(Array(data.length).fill(true))
  }
  setSelectedAll(!selectedAll)
}

function RatesTable({
  data = [],
  i18n
}){
  const [itemsClicked, setItemsClicked] = useState(Array(data.length).fill(false));
  const [selectedAll, setSelectedAll] = useState(false);
  const [ scrollableNode, setScrollableNode] = useState(null);
  const [offset, setOffset] = useState(RATE_COUNT);
  useEffect(()=>{
    setScrollableNode(window);
  }, [])
  return (
    <RatesTableWrapper>
      <Button onClick={()=>{
        onButtonClick({
          data,
          selectedAll,
          setItemsClicked,
          setSelectedAll
        })
      }}>
        {i18n["SELECT_ALL"]}
      </Button>
      <UlWrapper>
        <InfiniteScroll
          hasMore={data.length > offset}
          loader={<DelayedUiWrapper><div>...Loading</div></DelayedUiWrapper>}
          next={()=>{
            setOffset(offset + RATE_COUNT);
          }}
          scrollThreshold='80%'
          scrollableNode={scrollableNode}
        >
          {getItem({rates: data.slice(0, offset), itemsClicked, setItemsClicked})} 
        </InfiniteScroll>
      </UlWrapper>
    </RatesTableWrapper>
  )
}

export default withI18n(RatesTable)