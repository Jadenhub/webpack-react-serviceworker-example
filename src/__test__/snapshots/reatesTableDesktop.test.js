import renderer from 'react-test-renderer';
import RatesTable from '../../components/rates/RatesTable';
import React from 'react';
import { RATES } from '../mockData/data';
import { I18N } from '../mockData/i18n';
import { getTableData } from '../../components/hooks/useRateData';

test('renders correctly', () => {
  const data = getTableData(RATES)
  const tree = renderer
    .create(<RatesTable data={data} i18n={I18N}/>)
    .toJSON();
  expect(tree).toMatchSnapshot();
});

test('without data', () => {
  const data = getTableData(RATES)
  const tree = renderer
    .create(<RatesTable i18n={I18N}/>)
    .toJSON();
  expect(tree).toMatchSnapshot();
});