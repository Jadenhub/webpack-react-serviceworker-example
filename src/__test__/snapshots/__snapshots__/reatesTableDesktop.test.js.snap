// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
Array [
  <div
    className="RatesTable__PaginationWrapper-w8xhh9-0 jVDtaU"
  />,
  <table
    className="RatesTable__TableWrapper-w8xhh9-1 bZrfOn"
    role="table"
  >
    <thead>
      <tr
        role="row"
      >
        <th
          colSpan={1}
          role="columnheader"
        >
          <div>
            <input
              checked={false}
              onChange={[Function]}
              type="checkbox"
            />
          </div>
        </th>
        <th
          colSpan={1}
          role="columnheader"
        >
          Name
        </th>
        <th
          colSpan={1}
          role="columnheader"
        >
          Type
        </th>
        <th
          colSpan={1}
          role="columnheader"
        >
          Unit
        </th>
        <th
          colSpan={1}
          role="columnheader"
        >
          Value
        </th>
      </tr>
    </thead>
    <tbody
      role="rowgroup"
    >
      <tr
        role="row"
      >
        <td
          role="cell"
        >
          <div>
            <input
              checked={false}
              onChange={[Function]}
              type="checkbox"
            />
          </div>
        </td>
        <td
          role="cell"
        >
          
        </td>
        <td
          role="cell"
        >
          
        </td>
        <td
          role="cell"
        >
          
        </td>
        <td
          role="cell"
        >
          
        </td>
      </tr>
    </tbody>
  </table>,
]
`;

exports[`without data 1`] = `
Array [
  <div
    className="RatesTable__PaginationWrapper-w8xhh9-0 jVDtaU"
  />,
  <table
    className="RatesTable__TableWrapper-w8xhh9-1 bZrfOn"
    role="table"
  >
    <thead>
      <tr
        role="row"
      >
        <th
          colSpan={1}
          role="columnheader"
        >
          <div>
            <input
              checked={false}
              onChange={[Function]}
              type="checkbox"
            />
          </div>
        </th>
        <th
          colSpan={1}
          role="columnheader"
        >
          Name
        </th>
        <th
          colSpan={1}
          role="columnheader"
        >
          Type
        </th>
        <th
          colSpan={1}
          role="columnheader"
        >
          Unit
        </th>
        <th
          colSpan={1}
          role="columnheader"
        >
          Value
        </th>
      </tr>
    </thead>
    <tbody
      role="rowgroup"
    />
  </table>,
]
`;
